File: 121_Kth Largest Element in an Array.cpp
/*
121. Kth Largest Element in an Array
Find the k-th largest element in an unsorted array.
We'll use nth_element for average O(n) time.
*/
#include <bits/stdc++.h>
using namespace std;

int findKthLargest(vector<int>& nums, int k) {
    int n = nums.size();
    nth_element(nums.begin(), nums.begin() + (n - k), nums.end());
    return nums[n - k];
}

int main() {
    vector<int> a{3,2,1,5,6,4};
    cout << "kthLargest([3,2,1,5,6,4],2) = " << findKthLargest(a,2) << "\n"; // 5

    vector<int> b{3,2,3,1,2,4,5,5,6};
    cout << "kthLargest(...,4) = " << findKthLargest(b,4) << "\n"; // 4

    vector<int> c{1};
    cout << "kthLargest([1],1) = " << findKthLargest(c,1) << "\n"; // 1
    return 0;
}

// Time Complexity: Average O(n) (nth_element)
// Space Complexity: O(1)

File: 121_Kth Largest Element in an Array.py
"""
121. Kth Largest Element in an Array
Find the k-th largest element in an unsorted array. Use heapq.nlargest for simplicity.
"""
from typing import List
import heapq

def find_kth_largest(nums: List[int], k: int) -> int:
    # heapq.nlargest returns k largest elements in descending order
    return heapq.nlargest(k, nums)[-1]

if __name__ == '__main__':
    print('kthLargest([3,2,1,5,6,4],2) =', find_kth_largest([3,2,1,5,6,4],2))  # 5
    print('kthLargest([3,2,3,1,2,4,5,5,6],4) =', find_kth_largest([3,2,3,1,2,4,5,5,6],4))  # 4
    print('kthLargest([1],1) =', find_kth_largest([1],1))  # 1

# Time Complexity: O(n log k) or O(n) depending on implementation
# Space Complexity: O(k)
